File: swift/Sources/Fory/PrimitiveSerializers.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

extension Bool: Serializer {
    public static var staticTypeId: TypeId { .bool }

    public static func foryDefault() -> Bool { false }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeUInt8(self ? 1 : 0)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Bool {
        try context.buffer.readUInt8() != 0
    }
}

extension Int8: Serializer {
    public static var staticTypeId: TypeId { .int8 }

    public static func foryDefault() -> Int8 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeInt8(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Int8 {
        try context.buffer.readInt8()
    }
}

extension Int16: Serializer {
    public static var staticTypeId: TypeId { .int16 }

    public static func foryDefault() -> Int16 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeInt16(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Int16 {
        try context.buffer.readInt16()
    }
}

extension Int32: Serializer {
    public static var staticTypeId: TypeId { .varint32 }

    public static func foryDefault() -> Int32 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeVarInt32(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Int32 {
        try context.buffer.readVarInt32()
    }
}

extension Int64: Serializer {
    public static var staticTypeId: TypeId { .varint64 }

    public static func foryDefault() -> Int64 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeVarInt64(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Int64 {
        try context.buffer.readVarInt64()
    }
}

extension UInt8: Serializer {
    public static var staticTypeId: TypeId { .uint8 }

    public static func foryDefault() -> UInt8 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeUInt8(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> UInt8 {
        try context.buffer.readUInt8()
    }
}

extension UInt16: Serializer {
    public static var staticTypeId: TypeId { .uint16 }

    public static func foryDefault() -> UInt16 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeUInt16(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> UInt16 {
        try context.buffer.readUInt16()
    }
}

extension UInt32: Serializer {
    public static var staticTypeId: TypeId { .varUInt32 }

    public static func foryDefault() -> UInt32 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeVarUInt32(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> UInt32 {
        try context.buffer.readVarUInt32()
    }
}

extension UInt64: Serializer {
    public static var staticTypeId: TypeId { .varUInt64 }

    public static func foryDefault() -> UInt64 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeVarUInt64(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> UInt64 {
        try context.buffer.readVarUInt64()
    }
}

public struct ForyInt32Fixed: Serializer, Equatable {
    public var rawValue: Int32 = 0
    public init(rawValue: Int32 = 0) {
        self.rawValue = rawValue
    }

    public static func foryDefault() -> ForyInt32Fixed { .init() }
    public static var staticTypeId: TypeId { .int32 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeInt32(rawValue)
    }

    public static func foryReadData(_ context: ReadContext) throws -> ForyInt32Fixed {
        .init(rawValue: try context.buffer.readInt32())
    }
}

public struct ForyInt64Fixed: Serializer, Equatable {
    public var rawValue: Int64 = 0
    public init(rawValue: Int64 = 0) {
        self.rawValue = rawValue
    }

    public static func foryDefault() -> ForyInt64Fixed { .init() }
    public static var staticTypeId: TypeId { .int64 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeInt64(rawValue)
    }

    public static func foryReadData(_ context: ReadContext) throws -> ForyInt64Fixed {
        .init(rawValue: try context.buffer.readInt64())
    }
}

public struct ForyInt64Tagged: Serializer, Equatable {
    public var rawValue: Int64 = 0
    public init(rawValue: Int64 = 0) {
        self.rawValue = rawValue
    }

    public static func foryDefault() -> ForyInt64Tagged { .init() }
    public static var staticTypeId: TypeId { .taggedInt64 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeTaggedInt64(rawValue)
    }

    public static func foryReadData(_ context: ReadContext) throws -> ForyInt64Tagged {
        .init(rawValue: try context.buffer.readTaggedInt64())
    }
}

public struct ForyUInt32Fixed: Serializer, Equatable {
    public var rawValue: UInt32 = 0
    public init(rawValue: UInt32 = 0) {
        self.rawValue = rawValue
    }

    public static func foryDefault() -> ForyUInt32Fixed { .init() }
    public static var staticTypeId: TypeId { .uint32 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeUInt32(rawValue)
    }

    public static func foryReadData(_ context: ReadContext) throws -> ForyUInt32Fixed {
        .init(rawValue: try context.buffer.readUInt32())
    }
}

public struct ForyUInt64Fixed: Serializer, Equatable {
    public var rawValue: UInt64 = 0
    public init(rawValue: UInt64 = 0) {
        self.rawValue = rawValue
    }

    public static func foryDefault() -> ForyUInt64Fixed { .init() }
    public static var staticTypeId: TypeId { .uint64 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeUInt64(rawValue)
    }

    public static func foryReadData(_ context: ReadContext) throws -> ForyUInt64Fixed {
        .init(rawValue: try context.buffer.readUInt64())
    }
}

public struct ForyUInt64Tagged: Serializer, Equatable {
    public var rawValue: UInt64 = 0
    public init(rawValue: UInt64 = 0) {
        self.rawValue = rawValue
    }

    public static func foryDefault() -> ForyUInt64Tagged { .init() }
    public static var staticTypeId: TypeId { .taggedUInt64 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeTaggedUInt64(rawValue)
    }

    public static func foryReadData(_ context: ReadContext) throws -> ForyUInt64Tagged {
        .init(rawValue: try context.buffer.readTaggedUInt64())
    }
}

#if arch(arm64) || arch(x86_64)
extension Int: Serializer {
    public static var staticTypeId: TypeId { .varint64 }

    public static func foryDefault() -> Int { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeVarInt64(Int64(self))
    }

    public static func foryReadData(_ context: ReadContext) throws -> Int {
        Int(try context.buffer.readVarInt64())
    }
}

extension UInt: Serializer {
    public static var staticTypeId: TypeId { .varUInt64 }

    public static func foryDefault() -> UInt { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeVarUInt64(UInt64(self))
    }

    public static func foryReadData(_ context: ReadContext) throws -> UInt {
        UInt(try context.buffer.readVarUInt64())
    }
}
#endif

extension Float: Serializer {
    public static var staticTypeId: TypeId { .float32 }

    public static func foryDefault() -> Float { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeFloat32(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Float {
        try context.buffer.readFloat32()
    }
}

extension Double: Serializer {
    public static var staticTypeId: TypeId { .float64 }

    public static func foryDefault() -> Double { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeFloat64(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Double {
        try context.buffer.readFloat64()
    }
}

public struct BFloat16: Serializer, Equatable, Hashable, Sendable {
    public var rawValue: UInt16

    public init(rawValue: UInt16 = 0) {
        self.rawValue = rawValue
    }

    public static func foryDefault() -> BFloat16 { .init() }
    public static var staticTypeId: TypeId { .bfloat16 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeUInt16(rawValue)
    }

    public static func foryReadData(_ context: ReadContext) throws -> BFloat16 {
        .init(rawValue: try context.buffer.readUInt16())
    }
}

extension Float16: Serializer {
    public static var staticTypeId: TypeId { .float16 }

    public static func foryDefault() -> Float16 { 0 }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        _ = hasGenerics
        context.buffer.writeUInt16(bitPattern)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Float16 {
        Float16(bitPattern: try context.buffer.readUInt16())
    }
}

private enum StringEncoding: UInt64 {
    case latin1 = 0
    case utf16 = 1
    case utf8 = 2
}

private func decodeLatin1(_ bytes: [UInt8]) -> String {
    var scalarView = String.UnicodeScalarView()
    scalarView.reserveCapacity(bytes.count)
    for byte in bytes {
        scalarView.append(UnicodeScalar(UInt32(byte))!)
    }
    return String(scalarView)
}

extension String: Serializer {
    public static var staticTypeId: TypeId { .string }

    public static func foryDefault() -> String { "" }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readStaticTypeInfo(staticTypeId)
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        let utf8Bytes = self.utf8
        let header = (UInt64(utf8Bytes.count) << 2) | StringEncoding.utf8.rawValue
        if utf8Bytes.withContiguousStorageIfAvailable({ contiguousBytes in
            let totalBytes = UnsafeUtil.varUInt64Size(header) + contiguousBytes.count
            UnsafeUtil.writeRegion(buffer: context.buffer, exactCount: totalBytes) { base in
                var index = UnsafeUtil.writeVarUInt64(header, to: base, index: 0)
                index = UnsafeUtil.copyBytes(contiguousBytes, to: base, index: index)
                assert(index == totalBytes)
            }
            return true
        }) != nil {
            return
        }
        context.buffer.writeVarUInt36Small(header)
        context.buffer.writeBytes(utf8Bytes)
    }

    public static func foryReadData(_ context: ReadContext) throws -> String {
        let header = try context.buffer.readVarUInt36Small()
        let encoding = header & 0x03
        let byteLength = Int(header >> 2)

        switch encoding {
        case StringEncoding.utf8.rawValue:
            return try context.buffer.readUTF8String(count: byteLength)
        case StringEncoding.latin1.rawValue:
            let bytes = try context.buffer.readBytes(count: byteLength)
            return decodeLatin1(bytes)
        case StringEncoding.utf16.rawValue:
            let bytes = try context.buffer.readBytes(count: byteLength)
            if (byteLength & 1) != 0 {
                throw ForyError.encodingError("utf16 byte length is not even")
            }
            var units: [UInt16] = []
            units.reserveCapacity(byteLength / 2)
            var index = 0
            while index < bytes.count {
                let lo = UInt16(bytes[index])
                let hi = UInt16(bytes[index + 1]) << 8
                units.append(lo | hi)
                index += 2
            }
            return String(decoding: units, as: UTF16.self)
        default:
            throw ForyError.encodingError("unsupported string encoding \(encoding)")
        }
    }
}

extension Data: Serializer {
    public static var staticTypeId: TypeId { .binary }

    public static func foryDefault() -> Data { Data() }

    public static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        context.writeStaticTypeInfo(staticTypeId)
    }

    public static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        let rawTypeID = try context.buffer.readVarUInt32()
        guard let typeID = TypeId(rawValue: rawTypeID) else {
            throw ForyError.invalidData("unknown type id \(rawTypeID)")
        }
        if typeID != .binary && typeID != .uint8Array {
            throw ForyError.typeMismatch(expected: TypeId.binary.rawValue, actual: rawTypeID)
        }
        return nil
    }

    public func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        context.buffer.writeVarUInt32(UInt32(self.count))
        context.buffer.writeData(self)
    }

    public static func foryReadData(_ context: ReadContext) throws -> Data {
        let length = try context.buffer.readVarUInt32()
        let byteLength = Int(length)
        try context.ensureBinaryLength(byteLength, label: "binary")
        try context.ensureRemainingBytes(byteLength, label: "binary")
        let bytes = try context.buffer.readBytes(count: byteLength)
        return Data(bytes)
    }
}

File: java/fory-core/src/main/java/org/apache/fory/serializer/PrimitiveSerializers.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.serializer;

import static org.apache.fory.type.TypeUtils.PRIMITIVE_LONG_TYPE;

import org.apache.fory.context.ReadContext;
import org.apache.fory.context.WriteContext;

import org.apache.fory.Fory;
import org.apache.fory.config.Config;
import org.apache.fory.codegen.Expression;
import org.apache.fory.codegen.Expression.Invoke;
import org.apache.fory.config.LongEncoding;
import org.apache.fory.memory.MemoryBuffer;
import org.apache.fory.memory.Platform;
import org.apache.fory.resolver.TypeResolver;
import org.apache.fory.type.Float16;
import org.apache.fory.util.Preconditions;

/** Serializers for java primitive types. */
@SuppressWarnings({"rawtypes", "unchecked"})
public class PrimitiveSerializers {
  public static final class BooleanSerializer extends ImmutableSerializer<Boolean> {
    public BooleanSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Boolean value) {
      writeContext.getBuffer().writeBoolean(value);
    }

    @Override
    public Boolean read(ReadContext readContext) {
      return readContext.getBuffer().readBoolean();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class ByteSerializer extends ImmutableSerializer<Byte> {
    public ByteSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Byte value) {
      writeContext.getBuffer().writeByte(value);
    }

    @Override
    public Byte read(ReadContext readContext) {
      return readContext.getBuffer().readByte();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class Uint8Serializer extends ImmutableSerializer<Integer> {
    public Uint8Serializer(Config config) {
      super(config, Integer.class);
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      Preconditions.checkArgument(value >= 0 && value <= 255);
      writeContext.getBuffer().writeByte(value.byteValue());
    }

    @Override
    public Integer read(ReadContext readContext) {
      int b = readContext.getBuffer().readByte();
      return b >>> 24;
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class Uint16Serializer extends ImmutableSerializer<Integer> {
    public Uint16Serializer(Config config) {
      super(config, Integer.class);
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      Preconditions.checkArgument(value >= 0 && value <= 65535);
      writeContext.getBuffer().writeByte(value.byteValue());
    }

    @Override
    public Integer read(ReadContext readContext) {
      int b = readContext.getBuffer().readByte();
      return b >>> 16;
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class CharSerializer extends ImmutableSerializer<Character> {
    public CharSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Character value) {
      writeContext.getBuffer().writeChar(value);
    }

    @Override
    public Character read(ReadContext readContext) {
      return readContext.getBuffer().readChar();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class ShortSerializer extends ImmutableSerializer<Short> {
    public ShortSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Short value) {
      writeContext.getBuffer().writeInt16(value);
    }

    @Override
    public Short read(ReadContext readContext) {
      return readContext.getBuffer().readInt16();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class IntSerializer extends ImmutableSerializer<Integer> {
    private final boolean compressNumber;

    public IntSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
      // Cross-language encoding always uses varint; Java mode follows compressInt config.
      compressNumber = config.isXlang() || config.compressInt();
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      if (compressNumber) {
        writeContext.getBuffer().writeVarInt32(value);
      } else {
        writeContext.getBuffer().writeInt32(value);
      }
    }

    @Override
    public Integer read(ReadContext readContext) {
      if (compressNumber) {
        return readContext.getBuffer().readVarInt32();
      }
      return readContext.getBuffer().readInt32();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class VarUint32Serializer extends Serializer<Integer> {
    public VarUint32Serializer(Config config) {
      super(config, Integer.class);
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      Preconditions.checkArgument(value >= 0);
      writeContext.getBuffer().writeVarUint32(value);
    }

    @Override
    public Integer read(ReadContext readContext) {
      return readContext.getBuffer().readVarUint32();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class LongSerializer extends ImmutableSerializer<Long> {
    private final LongEncoding longEncoding;

    public LongSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
      longEncoding = config.isXlang() ? LongEncoding.VARINT : config.longEncoding();
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      writeInt64(writeContext.getBuffer(), value, longEncoding);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readInt64(readContext.getBuffer(), longEncoding);
    }

    public static Expression writeInt64(
        Expression buffer, Expression v, LongEncoding longEncoding, boolean ensureBounds) {
      switch (longEncoding) {
        case FIXED:
          return new Invoke(buffer, "writeInt64", v);
        case TAGGED:
          return new Invoke(
              buffer, ensureBounds ? "writeTaggedInt64" : "_unsafeWriteTaggedInt64", v);
        case VARINT:
          return new Invoke(buffer, ensureBounds ? "writeVarInt64" : "_unsafeWriteVarInt64", v);
        default:
          throw new UnsupportedOperationException("Unsupported long encoding " + longEncoding);
      }
    }

    public static void writeInt64(MemoryBuffer buffer, long value, LongEncoding longEncoding) {
      if (longEncoding == LongEncoding.TAGGED) {
        buffer.writeTaggedInt64(value);
      } else if (longEncoding == LongEncoding.FIXED) {
        buffer.writeInt64(value);
      } else {
        buffer.writeVarInt64(value);
      }
    }

    public static long readInt64(MemoryBuffer buffer, LongEncoding longEncoding) {
      if (longEncoding == LongEncoding.TAGGED) {
        return buffer.readTaggedInt64();
      } else if (longEncoding == LongEncoding.FIXED) {
        return buffer.readInt64();
      } else {
        return buffer.readVarInt64();
      }
    }

    public static Expression readInt64(Expression buffer, LongEncoding longEncoding) {
      return new Invoke(buffer, readLongFunc(longEncoding), PRIMITIVE_LONG_TYPE);
    }

    public static String readLongFunc(LongEncoding longEncoding) {
      switch (longEncoding) {
        case FIXED:
          return Platform.IS_LITTLE_ENDIAN ? "_readInt64OnLE" : "_readInt64OnBE";
        case TAGGED:
          return Platform.IS_LITTLE_ENDIAN ? "_readTaggedInt64OnLE" : "_readTaggedInt64OnBE";
        case VARINT:
          return Platform.IS_LITTLE_ENDIAN ? "_readVarInt64OnLE" : "_readVarInt64OnBE";
        default:
          throw new UnsupportedOperationException("Unsupported long encoding " + longEncoding);
      }
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class VarUint64Serializer extends Serializer<Long> {
    public VarUint64Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      Preconditions.checkArgument(value >= 0);
      writeContext.getBuffer().writeVarUint64(value);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readContext.getBuffer().readVarUint64();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class FloatSerializer extends ImmutableSerializer<Float> {
    public FloatSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Float value) {
      writeContext.getBuffer().writeFloat32(value);
    }

    @Override
    public Float read(ReadContext readContext) {
      return readContext.getBuffer().readFloat32();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class DoubleSerializer extends ImmutableSerializer<Double> {
    public DoubleSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Double value) {
      writeContext.getBuffer().writeFloat64(value);
    }

    @Override
    public Double read(ReadContext readContext) {
      return readContext.getBuffer().readFloat64();
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static final class Float16Serializer extends ImmutableSerializer<Float16> {
    public Float16Serializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Float16 value) {
      writeContext.getBuffer().writeInt16(value.toBits());
    }

    @Override
    public Float16 read(ReadContext readContext) {
      return Float16.fromBits(readContext.getBuffer().readInt16());
    }

    @Override
    public boolean threadSafe() {
      return true;
    }
  }

  public static void registerDefaultSerializers(TypeResolver resolver) {
    // primitive types will be boxed.
    Config config = resolver.getConfig();
    resolver.registerInternalSerializer(boolean.class, new BooleanSerializer(config, boolean.class));
    resolver.registerInternalSerializer(byte.class, new ByteSerializer(config, byte.class));
    resolver.registerInternalSerializer(short.class, new ShortSerializer(config, short.class));
    resolver.registerInternalSerializer(char.class, new CharSerializer(config, char.class));
    resolver.registerInternalSerializer(int.class, new IntSerializer(config, int.class));
    resolver.registerInternalSerializer(long.class, new LongSerializer(config, long.class));
    resolver.registerInternalSerializer(float.class, new FloatSerializer(config, float.class));
    resolver.registerInternalSerializer(double.class, new DoubleSerializer(config, double.class));
    resolver.registerInternalSerializer(Boolean.class, new BooleanSerializer(config, Boolean.class));
    resolver.registerInternalSerializer(Byte.class, new ByteSerializer(config, Byte.class));
    resolver.registerInternalSerializer(Short.class, new ShortSerializer(config, Short.class));
    resolver.registerInternalSerializer(
        Character.class, new CharSerializer(config, Character.class));
    resolver.registerInternalSerializer(Integer.class, new IntSerializer(config, Integer.class));
    resolver.registerInternalSerializer(Long.class, new LongSerializer(config, Long.class));
    resolver.registerInternalSerializer(Float.class, new FloatSerializer(config, Float.class));
    resolver.registerInternalSerializer(Double.class, new DoubleSerializer(config, Double.class));
    resolver.registerInternalSerializer(
        Float16.class, new Float16Serializer(config, Float16.class));
  }
}
